libnmstate/netapplier.py

```diff
@@ -59,20 +57,33 @@ def apply(
     :returns: Checkpoint identifier
     :rtype: str
     """
-    desired_state = copy.deepcopy(desired_state)
-    validator.validate(desired_state)
-    validator.validate_capabilities(desired_state, netinfo.capabilities())
-    validator.validate_unique_interface_name(desired_state)
-    validator.validate_dhcp(desired_state)
-    validator.validate_dns(desired_state)
-    validator.validate_vxlan(desired_state)
-    validator.validate_bridge(desired_state)
-
-    checkpoint = _apply_ifaces_state(
-        state.State(desired_state), verify_change, commit, rollback_timeout
-    )
-    if checkpoint:
-        return str(checkpoint.dbuspath)
+    with plugin_context() as plugin:
+        validator.validate(desired_state)
+        validator.validate_capabilities(desired_state, plugin.capabilities)
+        validator.validate_unique_interface_name(desired_state)
+        validator.validate_dhcp(desired_state)
+        validator.validate_dns(desired_state)
+        validator.validate_vxlan(desired_state)
+        validator.validate_bridge(desired_state)
+        checkpoint = plugin.create_checkpoint(rollback_timeout)
+        try:
+            _apply_ifaces_state(
+                plugin, state.State(desired_state), verify_change
+            )
+        except NmstateError:
+            checkpoint.rollback()
+            checkpoint = None
+            # Assume rollback occurred.
+            # Checkpoint rollback is async, there is a need to wait for it to
+            # finish before proceeding with other actions.
+            time.sleep(5)
```

Comment:
This is not an issue related with this PR. But maybe is there a way to use checkpoints instead of sleeping 5 seconds. It could improve the time. I will investigate it. It's on me, just writing it down to have it as a reminder and asking for your opinion.

Reply:
The reason we sleep 5 seconds:
we are using dbus interface directly for checkpoint stuff. Once we switch to libnm for checkpoint, we can wait the rollback been finished.

The PR for replacing dbus with libnm functions is: https://github.com/nmstate/nmstate/pull/236

I postponed that PR as we are making a lot changes to code bases, and this is trivial.

Reply:
NM does not really signal when rollback is completed. It has the same problem with semantics as "when is a device fully activated". Therefore, to do this correctly, Nmstate would need to wait until all devices affected by the checkpoint are in the expected activation state.

Reply:
Nice, thanks for explaining everything.